### PR TITLE
Drag and drop domUtil helpers

### PR DESCRIPTION
**Pull request by afshin**
These are three helper functions for allowing users to attach arbitrary data to `DragEvent` instances generated by the various drag and drop operations.

These methods are meant to supersede the native `event.dataTransfer.getData()` and `.setData()` methods, which can only use strings as data.

To see how they apply to our dashboarding scenario, the `dragdrop.html` example here is illustrative:
https://github.com/afshin/phosphor-splitpanel/commit/7e3c951005ccf20698bfb7a4d35e2b2e4c71a126

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -268,3 +268,47 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
+
+const DRAG_MIME_TYPE = 'application/x-phosphor-draggable';
+
+var dropCache: { data: { [reference: string]: any }, id: number } = {
+  data: { },
+  id: 0
+};
+
+/**
+ * Get arbitrary data that is associated with a drag and drop operation.
```
Comment (blink1073) [inline src/index.ts]: remove `that is`.

File: src/index.ts [hunk 2]
```diff
@@ -268,3 +268,47 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
+ *
+ * This method will typically be called in a drop event handler.
```
Comment (blink1073) [inline src/index.ts]: Missing a description of what it does (even if obvious)

File: src/index.ts (hunk 1 above)
Comment (afshin) [inline src/index.ts]: Concise. Good plan.

File: src/index.ts [hunk 3]
```diff
@@ -268,3 +268,47 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
+function getDropData(event: DragEvent): any {
+  let reference = event.dataTransfer.getData(DRAG_MIME_TYPE);
+  return dropCache.data[reference];
+}
+
+
+/**
+ * Set arbitrary data that is associated with a specific drag event.
+ *
+ * This method should be called in dragstart event handlers.
```
Comment (blink1073) [inline src/index.ts]: Another nit: remove `This method`

File: test/src/index.ts [hunk 4]
```diff
@@ -167,4 +168,40 @@ describe('phosphor-domutil', () => {
 
   });
 
+  (() => {
```
Comment (sccolbert) [inline test/src/index.ts]: There's no need for this self-executing function

Comment (afshin) [inline test/src/index.ts]: Okay! I was going for hygiene, but it's probably overkill in this context.

Comment (sccolbert) [inline test/src/index.ts]: Yep. Try to stick with the existing patterns. If you need to declare some static stuff at the top of the module, that's fine.

File: test/src/index.ts [hunk 5]
```diff
@@ -10,12 +10,33 @@
...
+    getData: (mime: string): string => {
+      return mime in data ? data[mime] : null;
+    },
+    setData: (mime: string, datum: string): void => {
+      data[mime] = datum;
+    }
+  };
+  return <DragEvent>event;
+}
+
```
Comment (sccolbert) [inline test/src/index.ts]: 2 blank lines

File: test/src/index.ts [hunk 6]
```diff
@@ -10,12 +10,33 @@
...
+var dragPayloadOne = () => { /* an arbitrary function */ };
+var dragMimeOne = 'application/x-phosphor-test-one';
+var dragPayloadTwo = { an: 'arbitrary', value: 0 };
+var dragMimeTwo = 'application/x-phosphor-test-two';
+var dragEvent = createDragEvent();
+
```
Comment (sccolbert) [inline test/src/index.ts]: 2 blank lines

File: test/src/index.ts [hunk 7]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
 
   });
 
+  describe('setDropData', () => {
+    it('should set arbitrary data attached to a DragEvent', () => {
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: test/src/index.ts [hunk 8]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
...
+      expect(setDropData(dragEvent, dragMimeOne, dragPayloadOne)).to.be(void 0);
+      expect(setDropData(dragEvent, dragMimeTwo, dragPayloadTwo)).to.be(void 0);
+    });
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: test/src/index.ts [hunk 9]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
...
+  });
+
+  describe('getDropData', () => {
+    it('should get arbitrary data attached to a DragEvent', () => {
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: test/src/index.ts [hunk 10]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
...
+      expect(getDropData(dragEvent, dragMimeOne)).to.be(dragPayloadOne);
+      expect(getDropData(dragEvent, dragMimeTwo)).to.be(dragPayloadTwo);
+    });
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: test/src/index.ts [hunk 11]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
...
+  });
+
+  describe('clearDropData', () => {
+    it('should clear data attached to a DragEvent', () => {
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: test/src/index.ts [hunk 12]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
...
+      expect(clearDropData(dragEvent)).to.be(void 0);
+      expect(getDropData(dragEvent, dragMimeOne)).to.be(void 0);
+      expect(getDropData(dragEvent, dragMimeTwo)).to.be(void 0);
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: src/index.ts [hunk 13]
```diff
@@ -268,3 +268,75 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
+
+
+/**
+ * Clear data being held privately for a specific drag event.
+ *
+ * @param event - The DragEvent generated by a user interaction.
+ *
+ * #### Notes
+ * The native drag and drop life cycle ends with a `dragend` event on the
+ * draggable node. The value associated with a particular MIME should be cleared
```
Comment (blink1073) [inline src/index.ts]: The last note about `a particular MIME` is confusing, since all of the data is cleared.

File: test/src/index.ts [hunk 14]
```diff
@@ -167,4 +188,26 @@ describe('phosphor-domutil', () => {
...
+    });
```
Comment (sccolbert) [inline test/src/index.ts]: 1 blank line

File: src/index.ts [hunk 15]
```diff
@@ -268,3 +268,75 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
+ * @param event - The DragEvent generated by a user interaction.
+ *
+ * @param mime - The MIME type whose data is being retrieved.
+ *
+ * @param data - The value being stored for a DragEvent/MIME combination.
+ *
+ * #### Notes
+ * The native drag and drop life cycle begins with a `dragstart` event on the
+ * draggable node. The value associated with a particular MIME should be set in
+ * this event handler.
```
Comment (sccolbert) [inline src/index.ts]: 1 blank at end of file

File: src/index.ts [hunk 16]
```diff
@@ -268,3 +268,75 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
 }
+
+
+const DRAG_MIME_TYPE = 'application/x-phosphor-draggable';
+
+var dropCache = {
+  data: Object.create(null),
+  id: 0
+};
+
```
Comment (sccolbert) [inline src/index.ts]: 2 blanks between module entities

File: src/index.ts [hunk 17]
```diff
@@ -268,3 +268,75 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
     maxHeight: parseInt(cstyle.maxHeight, 10) || Infinity,
   }
 }
+
+
+const DRAG_MIME_TYPE = 'application/x-phosphor-draggable';
+
```
Comment (sccolbert) [inline src/index.ts]: 2 blanks between module entities

File: src/index.ts (hunk 13 above)
Comment (afshin) [inline src/index.ts]: That's a mistake, thanks!

File: src/index.ts [hunk 18]
```diff
@@ -268,3 +268,75 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
...
+ * draggable node. The value associated with a particular MIME should be set in
+ * this event handler.
+ */
+export
+function setDropData(event: DragEvent, mime: string, data: any): void {
+  let id = event.dataTransfer.getData(DRAG_MIME_TYPE);
+  if (!id) {
+    id = `drag-${dropCache.id++}`;
+    event.dataTransfer.setData(DRAG_MIME_TYPE, id);
+    dropCache.data[id] = { };
```
Comment (sccolbert) [inline src/index.ts]: `Object.create(null)` - This protects against the user providing `toString` as a mime and getting back the `Object.prototype.toString` function.

File: src/index.ts [hunk 19]
```diff
@@ -268,3 +268,75 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
     maxHeight: parseInt(cstyle.maxHeight, 10) || Infinity,
   }
 }
+
+
+const DRAG_MIME_TYPE = 'application/x-phosphor-draggable';
```
Comment (sccolbert) [inline src/index.ts]: Let's make this `application/x-phosphor-drag-data`

File: src/index.ts (hunk 18 above)
Comment (afshin) [inline src/index.ts]: Good point, I'll update.

File: src/index.ts (hunk 15 above)
Comment (afshin) [inline src/index.ts]: Hm, that's not the EOF I think.

File: src/index.ts [hunk 20]
```diff
@@ -268,3 +268,77 @@ function sizeLimits(node: HTMLElement): ISizeLimits {
     maxHeight: parseInt(cstyle.maxHeight, 10) || Infinity,
   }
 }
+
+
+const DRAG_MIME_TYPE = 'application/x-phosphor-drag-id';
```
Comment (blink1073) [inline src/index.ts]: We had settled on `x-phosphor-internal-drag-data`

**Comment by sccolbert** [comment]
w00t!

**Pull request merged**